### ISSUE TITLE
Report Shows Same Dependencies With Different Actual Version in Current and Outdated Section

### ISSUE THREAD

**Issue by mroeppis** [ben-manes/gradle-versions-plugin#509]
Using the default plugin configuration this is what shows up in my report:

```text
The following dependencies are using the latest milestone version:
 - ch.qos.logback:logback-classic:none
 - ch.qos.logback:logback-core:none
 - com.fasterxml.jackson.core:jackson-annotations:none
 - ...

The following dependencies have later milestone versions:
 - ch.qos.logback:logback-classic [1.2.3 -> 1.3.0-alpha5]
     http://logback.qos.ch
 - ch.qos.logback:logback-core [1.2.3 -> 1.3.0-alpha5]
     http://logback.qos.ch
 - com.fasterxml.jackson.core:jackson-annotations [2.12.2 -> 2.12.3]
     http://github.com/FasterXML/jackson
 - ...
```

I'm using a BOM platform module with exactly one version for each dependency. That's why I wonder how the report can introduce such ambiguity.

Is there a special meaning about the `none` versioned dependencies? If not it would make the report easier to understand if it didn't have those. One way of excluding them would be in:

https://github.com/ben-manes/gradle-versions-plugin/blob/6f7b95d52cbbaa9023384b9c6fb84c54937f712a/src/main/groovy/com/github/benmanes/gradle/versions/updates/DependencyUpdatesReporter.groovy#L225-L228

where a custom filter could be applied to the passed `dependencies`.

---

I'm using version `0.38.0` of the plugin running with Gradle `6.7.1`

**Comment by ben-manes** [comment]
I believe `none` is used to pass it through when the version is absent, as the `version` string is required when resolving the query. This was prior to Gradle's BOM support where this occurred for file-based dependencies, e.g. `libs/foo.jar`. Perhaps now an empty version is supported by Gradle, but I have not tested it. (This plugin was originally written for Gradle 1.0)

I don't recall the code, but maybe the latest section could display the resolved version instead of the original. They would otherwise be the same, except for BOMs where it would show you what you are using. We could also strip off the "none" for file-based dependencies as it has no meaning for the user.

P.S. PRs are welcome! 🙂

**Comment by mroeppis** [comment]
If different Gradle versions treat an empty version string differently at resolution then it's probably not safe to remove the `none` keyword in general. But removing it from the report seems safe to me. That's why I'd prefer to stick with my idea from above.

@ben-manes : If you still think there is another solution then please go into details.

**Comment by ben-manes** [comment]
In a recent version we bumped the minimum version from `1.0` (when this plugin was written) to `v5.something`. If desirable, it should be okay if later enough versions change and we set a new minimum.

I've tried to avoid feature creep and treat this more like a unix utility that other can script around. That's helped since I mostly rely on generous contributors rather than add new features, and try to maintain it wrt compatibility. In cases where one wanted to modify the report, I've leaned on the [custom report format](https://github.com/ben-manes/gradle-versions-plugin#custom-report-format) feature instead. That avoids a slippery slope because its otherwise a never ending list of possible changes.

A filter might be nice, but we'd want to be careful about where it goes in the lifecycle. That might mean late enough to be useful in more cases or discovering multiple hooks are needed. As much as possible, ideally features should piggyback on Gradle's concepts rather than reinvent them (or become divergent when Gradle adds their own flavor). This API work would require some diligence.

I'm not against an ad hoc change to strip `:none` from the results within the `Resolver`. That would hide the pass through and be more inline to what devs would expect, as you noted.

**Comment by mroeppis** [comment]
I've digged deeper into this.

Dependencies with `null` version cannot be stripped from declared dependencies:

https://github.com/ben-manes/gradle-versions-plugin/blob/7f9c97c07a97571cbc7feb7d7efca3fa23528665/src/main/groovy/com/github/benmanes/gradle/versions/updates/Resolver.groovy#L260-L263

If I do this the amount of reported updatable dependencies is almost halved. As a result we cannot strip the `none` keyword from `Coordinate` version:

https://github.com/ben-manes/gradle-versions-plugin/blob/7f9c97c07a97571cbc7feb7d7efca3fa23528665/src/main/groovy/com/github/benmanes/gradle/versions/updates/Coordinate.groovy#L40

What's possible is filtering out `none` versioned coordinates in case of unresolved dependencies:

https://github.com/ben-manes/gradle-versions-plugin/blob/7f9c97c07a97571cbc7feb7d7efca3fa23528665/src/main/groovy/com/github/benmanes/gradle/versions/updates/Resolver.groovy#L286-L290

This must be done before putting them in the map. In my opinion this is the earliest possible stage in the lifecycle to alleviate the issue.
By checking if the declared dependencies contain the coordinate key to be added we could also eliminate the following filter that merely does the same thing:

https://github.com/ben-manes/gradle-versions-plugin/blob/7f9c97c07a97571cbc7feb7d7efca3fa23528665/src/main/groovy/com/github/benmanes/gradle/versions/updates/Resolver.groovy#L304

My proposed fix would make it necessary to adjust a test case:

https://github.com/ben-manes/gradle-versions-plugin/blob/7f9c97c07a97571cbc7feb7d7efca3fa23528665/src/test/groovy/com/github/benmanes/gradle/versions/DependencyUpdatesSpec.groovy#L237-L238

Honestly, I'm a little startled to see that a dependency without version is meant to be reported up-to-date though it can't actually be compared to any candidate with another version. I'd expect such a case to be categorized `unresolved` but that would be another issue.

Are there any other suggestions or concerns? Otherwise I'd like to make a PR.

**Comment by ben-manes** [comment]
> Honestly, I'm a little startled to see that a dependency without version is meant to be reported up-to-date though it can't actually be compared to any candidate with another version. I'd expect such a case to be categorized unresolved but that would be another issue.

This was originally done for local file dependencies, which are [not versioned](https://docs.gradle.org/current/userguide/declaring_dependencies.html#sec:versioning_file_dependencies). The idiom is to version in the file name, but Gradle won't know this. There is no way to know if an what the true coordinate are for an opaque jar file, so it is up to date given that it lives outside of the dependency management's version conflict resolution. In that world someone used an escape hatch and its seemed just as reasonable to mark it as up-to-date as unresolved.

> My proposed fix would make it necessary to adjust a test case:

Thanks. I'm fine with changing the test.

**Comment by mroeppis** [comment]
Just to make things clear: The dependency in that test case won't show up anywhere in the report. I personally can live with that since I don't think it's useful trying to check a library that can't be displayed using _G:A:V_.
@ben-manes : Are you with me here?

If I need to use file dependencies I apply something like `implementation files('../lib/mySecretLib.jar')`. That lib doesn't show up in the report either (...has nothing to do with my proposed fix). As I wrote before, covering such cases and categorizing them correctly is another issue. If you see any potential here I'd open a separate one so we can analyze the circumstances.

**One more benefit**
The fix I proposed has another positive effect on the report. Actually the headline of this issue targets only dependencies that appear more than once in the report. But there are others like

`org.slf4j:jul-to-slf4j:none`

that show up in the Up-To-Date section only. The fix changes them into

`org.slf4j:jul-to-slf4j:1.7.30` (... this special case is up-to-date since I reject alpha and beta versions)

So we'd have a version number for any up-to-date dependency.

**Comment by ben-manes** [comment]
Ideally we should show every dependency in the report. Less for users and more for us to debug with, since sometimes it's easier to understand a misclassification (e.g. due to a bad resolution strategy) than absence of any details. It's not a blocker, so if we already miss cases then no harm. If it is easy to retain then please do try, but otherwise everything you said sounds good to me.

**Comment by mroeppis** [comment]
There is a way to save such _versionless_ dependencies but I don't like it.

This would move the fix outside of the `Resolver` and inside `DependencyUpdates`:
https://github.com/ben-manes/gradle-versions-plugin/blob/9bef5cb55bee63f2e51b3456c1172102cf4fb593/src/main/groovy/com/github/benmanes/gradle/versions/updates/DependencyUpdates.groovy#L67-L70

As can be seen we go over the projects and then over configurations associated with them. I suggest we save `none` versioned dependencies that appear only once in the resulting set of `DependencyStatus`. By doing so any `none` that has a real version equivalent, be it from the same or any other configuration, will be dropped.

Sadly such versionless dependencies will be reported up-to-date. As I mentioned before I'd prefer unresolved but that would mean to construct an instance of `UnresolvedDependency` containing a problem (`Throwable`) although there never was a problem resolving the latest version of that dependency (...due to the way query dependencies are built). That being said the clean way to go would be to create a new result category, e.g. _undeclared version_ containing dependencies (not constraints) that were explicitly declared without version by the user.

I guess this won't happen because it requires a bigger change (version mapping, reporting part). But bear in mind that saying "depX is up-to-date" is simply wrong if you can't tell the version of depX. This isn't about file dependencies as well since there is no reference to a file. Try declaring a dependency `implementation G:A` only. Gradle won't complain when configuring the project. But as soon as you try to compile it says:

```
Execution failed for task ':compileGroovy'.
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not find G:A:.
     Caused by: org.gradle.internal.resolve.ModuleVersionNotFoundException: Could not find G:A:.
```

This might be something to come back to in the near future. As for now it might be acceptable to leave the reporting as is and to change `DependencyUpdates` only.

What do you think?

**Comment by ben-manes** [comment]
I agree, `undeclared version` is the nicest idea. The reports are fairly simple: json/xml are automatic and the text/html are very simple. You might prefer to refactor to that, even though it is a bit more work.

I'm fine with whichever of the approaches that you decide on, as all make sense and have their code / effort impact. You're welcome to choose the path that best suits your time and preference. Thanks for digging into the details and thinking through the tradeoffs involved.
